File: tests/sdk/nodejs/new-resources.test.ts
/**
 * Mock Tests for newly added CAST AI resources and data sources (TypeScript).
 *
 * Verifies that the new resource and data source classes are exported from
 * the SDK and can be constructed. These are smoke tests only - they do not
 * exercise cloud-side behaviour.
 */

import * as pulumi from "@pulumi/pulumi";
import * as castai from "@castai/pulumi";
import { promisify } from "./test-utils";

/**
 * Default mock implementation that accepts every CAST AI token and returns
 * a deterministic id / pass-through state.
 */
class NewResourceMocks implements pulumi.runtime.Mocks {
    newResource(args: pulumi.runtime.MockResourceArgs): { id: string; state: any } {
        const outputs: any = { ...(args.inputs as any) };
        outputs.id = `${args.name}-id`;
        return { id: `${args.name}-id`, state: outputs };
    }

    call(args: pulumi.runtime.MockCallArgs): Record<string, any> {
        return {};
    }
}

pulumi.runtime.setMocks(new NewResourceMocks());

describe("New Resources", () => {
    it("exposes CacheGroup as a constructable resource", async () => {
        expect(castai.CacheGroup).toBeDefined();
        const res = new castai.CacheGroup("test-cache-group", {
            name: "test-cache-group",
            protocolType: "MySQL",
        });
        const id = await promisify(res.id);
        expect(id).toBeDefined();
    });

    it("exposes CacheConfiguration as a constructable resource", async () => {
        expect(castai.CacheConfiguration).toBeDefined();
        const res = new castai.CacheConfiguration("test-cache-config", {
            cacheGroupId: "cg-123",
            databaseName: "appdb",
            mode: "Auto",
        });
        const id = await promisify(res.id);
        expect(id).toBeDefined();
    });

    it("exposes CacheRule as a constructable resource", async () => {
        expect(castai.CacheRule).toBeDefined();
        const res = new castai.CacheRule("test-cache-rule", {
            cacheConfigurationId: "cc-123",
            cacheGroupId: "cg-123",
            mode: "Auto",
        });
        const id = await promisify(res.id);
        expect(id).toBeDefined();
    });

    it("exposes AiOptimizerModelRegistry as a constructable resource", async () => {
        expect(castai.AiOptimizerModelRegistry).toBeDefined();
        const res = new castai.AiOptimizerModelRegistry("test-model-registry", {
            credentials: "mock-credentials",
        });
        const id = await promisify(res.id);
        expect(id).toBeDefined();
    });

    it("exposes AiOptimizerModelSpecs as a constructable resource", async () => {
        expect(castai.AiOptimizerModelSpecs).toBeDefined();
        const res = new castai.AiOptimizerModelSpecs("test-model-specs", {
            model: "my-model",
            registryType: "public",
        });
        const id = await promisify(res.id);
        expect(id).toBeDefined();
    });

    it("exposes AiOptimizerHostedModel as a constructable resource", async () => {
        expect(castai.AiOptimizerHostedModel).toBeDefined();
        const res = new castai.AiOptimizerHostedModel("test-hosted-model", {
            clusterId: "cluster-123",
            modelSpecsId: "ms-123",
            port: 8080,
            service: "svc-123",
        });
        const id = await promisify(res.id);
        expect(id).toBeDefined();
    });

    it("exposes EnterpriseServiceAccount as a constructable resource", async () => {
        expect(castai.EnterpriseServiceAccount).toBeDefined();
        const res = new castai.EnterpriseServiceAccount("test-enterprise-sa", {
            enterpriseId: "ent-123",
        });
        const id = await promisify(res.id);
        expect(id).toBeDefined();
    });

    it("exposes WorkloadCustomMetricsDataSource as a constructable resource", async () => {
        expect(castai.WorkloadCustomMetricsDataSource).toBeDefined();
        const res = new castai.WorkloadCustomMetricsDataSource("test-workload-cmd", {
            clusterId: "cluster-123",
            prometheus: { url: "http://prometheus:9090" },
        });
        const id = await promisify(res.id);
        expect(id).toBeDefined();
    });

    it("exposes PodMutation as a constructable resource", async () => {
        expect(castai.PodMutation).toBeDefined();
        const res = new castai.PodMutation("test-pod-mutation", {
            clusterId: "cluster-123",
            enabled: true,
            filterV2: {},
        });
        const id = await promisify(res.id);
        expect(id).toBeDefined();
    });
});

describe("New Data Sources", () => {
    it("exposes getWorkloadScalingPolicies as a callable function", () => {
        expect(castai.getWorkloadScalingPolicies).toBeDefined();
        expect(typeof castai.getWorkloadScalingPolicies).toBe("function");
    });

    it("exposes getCacheGroup as a callable function", () => {
        expect(castai.getCacheGroup).toBeDefined();
        expect(typeof castai.getCacheGroup).toBe("function");
    });

    it("exposes getImpersonationServiceAccount as a callable function", () => {
        expect(castai.getImpersonationServiceAccount).toBeDefined();
        expect(typeof castai.getImpersonationServiceAccount).toBe("function");
    });
});

File: tests/sdk/go/new_resources_test.go
package tests

import (
	"testing"

	"github.com/castai/pulumi-castai/sdk/go/castai"
	"github.com/castai/pulumi-castai/sdk/go/castai/workload"
	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
	"github.com/stretchr/testify/assert"
)

// CastAINewResourceMocks is a permissive mock that returns a deterministic id
// for every resource type. Data source invocations return minimal values.
type CastAINewResourceMocks struct {
 pulumi.MockResourceMonitor
}

func (m *CastAINewResourceMocks) NewResource(args pulumi.MockResourceArgs) (string, resource.PropertyMap, error) {
 outputs := args.Inputs.Copy()
 id := args.Name + "-id"
 outputs["id"] = resource.NewStringProperty(id)
 return id, outputs, nil
}

func (m *CastAINewResourceMocks) Call(args pulumi.MockCallArgs) (resource.PropertyMap, error) {
 return resource.PropertyMap{}, nil
}

// RunNewResource is a helper that runs a Pulumi program with the new-resource
// mock monitor so that resource registration succeeds.
func runNewResource(t *testing.T, program func(ctx *pulumi.Context) error) {
 t.Helper()
 err := pulumi.RunErr(program, pulumi.WithMocks("project", "stack", &CastAINewResourceMocks{}))
 assert.NoError(t, err)
}

func TestCacheGroupConstructable(t *testing.T) {
 runNewResource(t, func(ctx *pulumi.Context) error {
 res, err := castai.NewCacheGroup(ctx, "test-cache-group", &castai.CacheGroupArgs{
 ProtocolType: pulumi.String("MySQL"),
 })
 assert.NoError(t, err)
 assert.NotNil(t, res)
 return nil
 })
}

func TestCacheConfigurationConstructable(t *testing.T) {
 runNewResource(t, func(ctx *pulumi.Context) error {
 res, err := castai.NewCacheConfiguration(ctx, "test-cache-config", &castai.CacheConfigurationArgs{
 CacheGroupId: pulumi.String("cg-123"),
 DatabaseName: pulumi.String("appdb"),
 Mode: pulumi.String("Auto"),
 })
 assert.NoError(t, err)
 assert.NotNil(t, res)
 return nil
 })
}

func TestCacheRuleConstructable(t *testing.T) {
 runNewResource(t, func(ctx *pulumi.Context) error {
 res, err := castai.NewCacheRule(ctx, "test-cache-rule", &castai.CacheRuleArgs{
 CacheConfigurationId: pulumi.String("cc-123"),
 CacheGroupId: pulumi.String("cg-123"),
 Mode: pulumi.String("Auto"),
 })
 assert.NoError(t, err)
 assert.NotNil(t, res)
 return nil
 })
}

func TestAiOptimizerModelRegistryConstructable(t *testing.T) {
 runNewResource(t, func(ctx *pulumi.Context) error {
 res, err := castai.NewAiOptimizerModelRegistry(ctx, "test-model-registry", &castai.AiOptimizerModelRegistryArgs{
 Credentials: pulumi.String("mock-credentials"),
 })
 assert.NoError(t, err)
 assert.NotNil(t, res)
 return nil
 })
}

func TestAiOptimizerModelSpecsConstructable(t *testing.T) {
 runNewResource(t, func(ctx *pulumi.Context) error {
 res, err := castai.NewAiOptimizerModelSpecs(ctx, "test-model-specs", &castai.AiOptimizerModelSpecsArgs{
 Model: pulumi.String("my-model"),
 RegistryType: pulumi.String("public"),
 })
 assert.NoError(t, err)
 assert.NotNil(t, res)
 return nil
 })
}

func TestAiOptimizerHostedModelConstructable(t *testing.T) {
 runNewResource(t, func(ctx *pulumi.Context) error {
 res, err := castai.NewAiOptimizerHostedModel(ctx, "test-hosted-model", &castai.AiOptimizerHostedModelArgs{
 ClusterId: pulumi.String("cluster-123"),
 ModelSpecsId: pulumi.String("ms-123"),
 Port: pulumi.Int(8080),
 Service: pulumi.String("svc-123"),
 })
 assert.NoError(t, err)
 assert.NotNil(t, res)
 return nil
 })
}

func TestEnterpriseServiceAccountConstructable(t *testing.T) {
 runNewResource(t, func(ctx *pulumi.Context) error {
 res, err := castai.NewEnterpriseServiceAccount(ctx, "test-enterprise-sa", &castai.EnterpriseServiceAccountArgs{
 EnterpriseId: pulumi.String("ent-123"),
 })
 assert.NoError(t, err)
 assert.NotNil(t, res)
 return nil
 })
}

func TestWorkloadCustomMetricsDataSourceConstructable(t *testing.T) {
 runNewResource(t, func(ctx *pulumi.Context) error {
 res, err := castai.NewWorkloadCustomMetricsDataSource(ctx, "test-workload-cmd", &castai.WorkloadCustomMetricsDataSourceArgs{
 ClusterId: pulumi.String("cluster-123"),
 Prometheus: &workload.WorkloadCustomMetricsDataSourcePrometheusArgs{
 Url: pulumi.String("http://prometheus:9090"),
 },
 })
 assert.NoError(t, err)
 assert.NotNil(t, res)
 return nil
 })
}

func TestPodMutationConstructable(t *testing.T) {
 runNewResource(t, func(ctx *pulumi.Context) error {
 res, err := castai.NewPodMutation(ctx, "test-pod-mutation", &castai.PodMutationArgs{
 ClusterId: pulumi.String("cluster-123"),
 Enabled: pulumi.Bool(true),
 FilterV2: &castai.PodMutationFilterV2Args{},
 })
 assert.NoError(t, err)
 assert.NotNil(t, res)
 return nil
 })
}
